src/twister2/builder/build_manager.py

```diff
@@ -0,0 +1,132 @@
+
+logger = logging.getLogger(__name__)
+
+
+class BuildStatus:
+    NOT_DONE = 0
+    IN_PROGRESS = 1
+    DONE = 2
+    FAILED = 3
+
+
+class BuildManager:
+    """
+    Class handles information about already built sources.
+
+    It allows to skip building when it was already built for another test.
+    """
+    _lock: FileLock = FileLock(BUILD_LOCK_FILE_PATH, timeout=1)
+
+    def __init__(self, output_dir: str | Path) -> None:
+        self._status_file: Path = Path(output_dir) / BUILD_STATUS_FILE_NAME
+        self.wait_timeout: int = 10 * 60  # seconds
+        self.initialize()
+
+    def initialize(self):
+        logger.info('Cleaning builder status file: %s', self._status_file)
+        with self._lock:
+            if self._status_file.exists():
+                return
+            self._write_data({})
+
+    def get_status(self, build_dir: str) -> int:
+        """
+        Return status for build source.
+
+        :param build_dir: path to build director
+        :return: build status
+        """
+        with self._lock:
+            data = self._read_data()
+            return data.get(str(build_dir), BuildStatus.NOT_DONE)
+
+    def _read_data(self):
+        with self._status_file.open(encoding='UTF-8') as file:
+            data: dict = json.load(file)
+        return data
+
+    def update_status(self, build_dir: str, status: int) -> bool:
+        """
+        Update status for build sorce.
+
+        If new status is equal to old one than return False,
+        otherwise return True
+
+        :param build_dir: path to build director
+        :param status: new status
+        :return: True if status was updated otherwise return False
+        """
+        with self._lock:
+            data = self._read_data()
+            if data.get(build_dir) == status:
+                return False
+            data[str(build_dir)] = status
+            self._write_data(data)
+            return True
+
+    def _write_data(self, data):
+        with self._status_file.open('w', encoding='UTF-8') as file:
+            json.dump(data, file, indent=2)
+
+    def build(self, builder: BuilderAbstract, build_config: BuildConfig) -> None:
+        """
+        Build source code.
+
+        :param builder: instance of a builder class
+        :param build_config: build configuration
+        """
+        status = self.get_status(build_config.build_dir)
+        if status == BuildStatus.DONE:
+            logger.info('Already build in %s', build_config.build_dir)
+            return
+        if status == BuildStatus.NOT_DONE:
+            if self.update_status(build_config.build_dir, BuildStatus.IN_PROGRESS) is True:
+                self._build(builder=builder, build_config=build_config)
+                return
+        if status == BuildStatus.IN_PROGRESS:
+            # another builder is building the same source
+            self._wait_for_build_to_finish(build_config.build_dir)
+            return
```

Comment:
This `return` can be removed. Thanks to this, there is no need to verify if build failed inside `_wait_for_build_to_finish` method.

Reply:
Actually it won't work because in the last `if` we check only if `status == BuildStatus.FAILED` (and we do not read status again). So if waiting for another building was broken due to timeout we will have still `IN_PROGRESS` status. It is better to failed test immediately while timeout appears.